Add baroclinic gyre tests (#547)
MPAS-Dev/compass: docs/users_guide/ocean/test_groups/baroclinic_gyre.rst

.. _baroclinic_gyre:

baroclinic_gyre
=======================

The ``baroclinic_gyre`` test group implements variants of the
baroclinic ocean gyre set-up from the 

> `MITgcm test case <https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html>`_.

Reviewer: I think sphinx is getting confused by having this same link twice.  I would suggest putting the following at the bottom of the page:
```
.. _MITgcm test case: https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html
```
And then take the URL out of this and the following link:
```suggestion
`MITgcm test case`_.
```

Author: Thanks for the suggestion, I modified accordingly. The documentation built fine for me, and the link was correct.

This test simulates a baroclinic, wind and buoyancy-forced, double-gyre ocean circulation. 
The grid employs spherical polar coordinates with 15 vertical layers.
The configuration is similar to the double-gyre setup first solved numerically
in `Cox and Bryan (1984) <https://journals.ametsoc.org/view/journals/phoc/14/4/1520-0485_1984_014_0674_anmotv_2_0_co_2.xml>`_: the model is configured to
represent an enclosed sector of fluid on a sphere, spanning the tropics to mid-latitudes,
:math:`60^{\circ} \times 60^{\circ}` in lateral extent.
The fluid is :math:`1.8`\ km deep and is forced by a zonal wind
stress which is constant in time, :math:`\tau_{\lambda}`, varying sinusoidally in the
north-south direction.

.. figure:: ../images/baroclinic_gyre_config.png
  :width: 95%
  :align: center
  :alt: baroclinic gyre configuration
  :name: baroclinic_gyre_config

>   Schematic of simulation domain and wind-stress forcing function for baroclinic gyre numerical experiment. The domain is enclosed by solid walls. From `MITgcm test case <https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html>`_.

Reviewer: After moving the URL to the bottom of the page as mentioned above:
```suggestion
  Schematic of simulation domain and wind-stress forcing function for baroclinic gyre numerical experiment. The domain is enclosed by solid walls. From `MITgcm test case`_.
```

.. _baroclinic_gyre:

> baroclinic_gyre

Reviewer: This needs to be added to the index.rst for test groups, just as in the developer's guide.

Author: @xylar I included that change in docs/users_guide/ocean/test_groups/index.rst. Am I missing something?

Reviewer: Sorry, I saw that and thought I deleted this comment. Please ignore it.

===============

The ``baroclinic_gyre`` test group implements variants of the
baroclinic ocean gyre set-up from the 
`MITgcm test case`_.

This test simulates a baroclinic, wind and buoyancy-forced, double-gyre ocean circulation. 
The grid employs spherical polar coordinates with 15 vertical layers.
The configuration is similar to the double-gyre setup first solved numerically
in `Cox and Bryan (1984) <https://journals.ametsoc.org/view/journals/phoc/14/4/1520-0485_1984_014_0674_anmotv_2_0_co_2.xml>`_: the model is configured to
represent an enclosed sector of fluid on a sphere, spanning the tropics to mid-latitudes,
:math:`60^{\circ} \times 60^{\circ}` in lateral extent.
The fluid is :math:`1.8`\ km deep and is forced by a zonal wind
stress which is constant in time, :math:`\tau_{\lambda}`, varying sinusoidally in the
north-south direction.

> .. figure:: ./images/baroclinic_gyre_config.png

Reviewer: We typically limit images to 500 pixels along the longer size to keep the repo small.  This file isn't super large but what do you think about shrinking it by a factor of 4?

.. math::
      \tau_{\lambda}(\varphi) = -\tau_{0}\cos \left(2 \pi \frac{\varphi-\varphi_o}{L_{\varphi}} \right)

where :math:`L_{\varphi}` is the lateral domain extent
(:math:`60^{\circ}`), :math:`\varphi_o` is set to :math:`15^{\circ} \text{N}` and :math:`\tau_0` is :math:`0.1 \text{ N m}^{-2}`.
:ref:`baroclinic_gyre.cfg` summarizes the
configuration options used in this simulation. 

Temperature is restored in the surface layer to a linear profile:

.. math::
      {\cal F}_\theta = - U_{piston} (\theta-\theta^*), \phantom{WWW}
   \theta^* = \frac{\theta_{\rm max} - \theta_{\rm min}}{L_\varphi} (\varphi_{\rm max} - \varphi) + \theta_{\rm min}
   :label: baroc_restore_theta

where the piston velocity :math:`U_{piston}` (in m.s^{-1}) is calculated by applying a relaxation timescale of 30 days (set in config file) over the thickness of the top layer (50 m by default) and :math:`\theta_{\rm max}=30^{\circ}` C, :math:`\theta_{\rm min}=0^{\circ}` C.

Initial state
-------------

Initially the fluid is stratified
with a reference potential temperature profile that varies from (approximately) :math:`\theta=30.6 \text{ } ^{\circ}`\ C

> in the surface layer to :math:`\theta=1.56 \text{ } ^{\circ}`\ C in the bottom layer. To ensure that the profile is independent of the vertical discretization, the profile is now set by a surface value (at the top interface) and a bottom value (at the bottom interface), set in the `.cfg` file. The default values have been chosen for the layer values (calculated with `zMid`) to approximate the discrete values presented in the MITgcm test case. The temperature functional form (and inner parameter `cc`  was determined by fitting an analytical function to the MITgcm discrete layer values (originally ranging from 2 to :math:`30 \text{ } ^{\circ}`\ C. If the `bottom_depth` is different from the default 1800m value, the temperature profile is stretched in the vertical to fit the surface and bottom temperature constraints, but the thermocline depth and the discrete layer values will move away from the MITgcm test case.  

Reviewer: ```suggestion
in the surface layer to :math:`\theta=1.56 \text{ } ^{\circ}`\ C in the bottom layer. To ensure that the profile is independent of the vertical discretization, the profile is now set by a surface value (at the top interface) and a bottom value (at the bottom interface), set in the ``.cfg`` file. The default values have been chosen for the layer values (calculated with ``zMid``) to approximate the discrete values presented in the MITgcm test case. The temperature functional form (and inner parameter ``cc``  was determined by fitting an analytical function to the MITgcm discrete layer values (originally ranging from 2 to :math:`30 \text{ } ^{\circ}`\ C. If the ``bottom_depth`` is different from the default 1800m value, the temperature profile is stretched in the vertical to fit the surface and bottom temperature constraints, but the thermocline depth and the discrete layer values will move away from the MITgcm test case.  
```

Reviewer: Maybe break up these long lines?

    
    # Total water column depth
    bottom_depth = 1800.
    
    # The type of vertical coordinate (e.g. z-level, z-star)
    coord_type = z-star
    
    # Whether to use "partial" or "full", or "None" to not alter the topography
    partial_cell_type = None
    
    # The minimum fraction of a layer for partial cells
    min_pc_fraction = 0.1
    
    
    # config options for MITgcm baroclinic gyre
    [baroclinic_gyre]
    # comment
    lat_min = 15
    lat_max = 75
    lon_min = 0
    lon_max = 60
    
    # Initial vertical temperature profile
    initial_temp_top = 33.
    initial_temp_bot = 1.
    
    # Constant salinity value (also used in restoring)
    initial_salinity = 34.
    
    # Maximum zonal wind stress value
    wind_stress_max = 0.1
    
    # Surface temperature restoring
    restoring_temp_min = 0.
    restoring_temp_max = 30.

    # Restoring timescale for surface temperature (in days)
    restoring_temp_timescale = 30.

performance_test
----------------

``ocean/baroclinic_gyre/performance_test`` is the default version of the
baroclinic_gyre test case for a short (3 time steps) test run and validation of
prognostic variables for regression testing.

> 3_year_test

Reviewer: Somewhere here (or above), I would say that a user can run the `forward` step several times, advancing the simulation by 3 years each time.  Or that they can alter the namelist file to increase the duration of the simulation.

Reviewer: Note however that you would need to make some changes I suggested to `forward.py` for this to actually work.

-----------

``ocean/baroclinic_gyre/3_year_test`` is an additional version of the
baroclinic_gyre test case with a longer (3-year) spin-up. By default, it 
includes monthly mean output, and plots the mean state of the simulation for the last 1 year (option in the config file). 

> Note that for the 80km configuration, the estimated time to equilibration is roughly 50 years (approx 3hours on default layout).  

Reviewer: ```suggestion
Note that for the 80km configuration, the estimated time to equilibration is roughly 50 years (approx 3 hours of compute time on default layout).  
```
